### PR TITLE
Allow broadcasting of value x params in Categorical

### LINKED ISSUE #79: Support broadcasting of sample x param
@alicanb observed that `Normal` supports broadcasting whereas `Categorical` does not:
```py
p = Variable(torch.Tensor([[0.1, 0.2, 0.3],[0.2, 0.2, 0.3]]), requires_grad=False)
dist = torch.distributions.Categorical(probs=p)
log_prob = dist.log_prob(dist.sample().unsqueeze(-1))
#RuntimeError: The expanded size of the tensor (1) must match the existing size (2) at non-singleton dimension 1
```

### PR DESCRIPTION
**Pull request by neerajprad**
Allows for broadcasting of values and parameters across the batch dimensions for Categorical. The other distributions do not need to do explicit broadcasting in `log_prob` and already handle this case.

Fixes #79.

### CODE REVIEW

File: torch/distributions/categorical.py [hunk 1]
```diff
@@ -67,7 +67,12 @@ def sample(self, sample_shape=torch.Size()):
 
     def log_prob(self, value):
         self._validate_log_prob_arg(value)
-        param_shape = value.size() + self.probs.size()[-1:]
+        if self.batch_shape:
```
Comment (fritzo) [inline torch/distributions/categorical.py]: I believe `self.batch_shape == self.probs.size()[:-1]` but is cheaper to compute if `logits` was specified. Consider replacing this with
```py
batch_shape = self.batch_shape
value_shape = torch._C.infer_size(value.shape, batch_shape) if batch_shape else value.shape
param_shape = value_shape + self._dim
```
and defining `self._dim` in the `__init__` method:
```py
self._dim = self._logits.shape[-1:] if self._probs is None else self._probs[-1:]
```

**Comment by fritzo** [comment]
Thanks for fixing this!

File: torch/distributions/categorical.py (hunk 1 above)
Comment (fritzo) [reply torch/distributions/categorical.py]: Or since you definitely need logits below, simply define logits as a local variable and query it for shape.

**Comment by neerajprad** [comment]
Thanks for the review, @fritzo. Addressed your comment.

**Comment by neerajprad** [comment]
Closed, and opened up https://github.com/pytorch/pytorch/pull/4614.

**Pull request closed without merging**